### PR TITLE
delete-unExisting-account

### PR DESCRIPTION
**Pull request by sasurobert**
Un bond waiting before switch

### CODE REVIEW

File: epochStart/metachain/systemSCs.go [hunk 1]
```diff
@@ -417,7 +417,17 @@ func (s *systemSCProcessor) fillStakingDataForNonEligible(validatorInfos map[uin
 
 			err := s.stakingDataProvider.FillValidatorInfo(validatorInfo.PublicKey)
 			if err != nil {
-				return err
+				log.Error("fillStakingDataForNonEligible", "error", err)
+				if validatorInfo.List == string(core.WaitingList) || validatorInfo.List == string(core.JailedList) {
+					return err
+				}
+
```
Comment (iulianpascalau) [inline epochStart/metachain/systemSCs.go:424]: I would add a log.Error here as this should not happen and we should correct this.

File: vm/systemSmartContracts/staking.go [hunk 2]
```diff
@@ -1462,6 +1462,10 @@ func (s *stakingSC) getOwner(args *vmcommon.ContractCallInput) vmcommon.ReturnCo
 		s.eei.AddReturnMessage(errGet.Error())
 		return vmcommon.UserError
 	}
+	if len(stakedData.OwnerAddress) == 0 {
+		s.eei.AddReturnMessage("owner address is nil")
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1466]: :+1: early warning.

File: epochStart/metachain/systemSCs.go (hunk 1 above)
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:424]: log . error is on line 420

Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:424]: log.error is on line 420

**Review by iulianpascalau — DISMISSED** [review]
integration test pending

**Review by iulianpascalau — DISMISSED** [review]

File: cmd/node/factory/structs.go [hunk 3]
```diff
@@ -2075,6 +2075,7 @@ func newMetaBlockProcessor(
 		RatingsData:      ratingsData,
 		EpochNotifier:    epochNotifier,
 		StakeEnableEpoch: systemSCConfig.StakingSystemSCConfig.StakeEnableEpoch,
+		UnBondCorrection: generalConfig.GeneralSettings.SCDeployEnableEpoch,
```
Comment (raduchis) [inline cmd/node/factory/structs.go:2078]: could be UnBondCorrection*Epoch*

File: process/scToProtocol/stakingToPeer.go [hunk 4]
```diff
@@ -47,13 +48,15 @@ type stakingToPeer struct {
...
-	jailRating       uint32
-	stakeEnableEpoch uint32
-	flagStaking      atomic.Flag
+	argParser             process.ArgumentsParser
+	currTxs               dataRetriever.TransactionCacher
+	startRating           uint32
+	unJailRating          uint32
+	jailRating            uint32
+	stakeEnableEpoch      uint32
+	unBondCorrectionEpoch uint32
```
Comment (raduchis) [inline process/scToProtocol/stakingToPeer.go:57]: :+1: for epoch

**Review by raduchis — DISMISSED** [review]

File: process/scToProtocol/stakingToPeer.go [hunk 5]
```diff
@@ -408,6 +415,9 @@ func (stp *stakingToPeer) getAllModifiedStates(body *block.Body) ([]string, erro
 func (stp *stakingToPeer) EpochConfirmed(epoch uint32) {
 	stp.flagStaking.Toggle(epoch >= stp.stakeEnableEpoch)
 	log.Debug("stakingToPeer: stake", "enabled", stp.flagStaking.IsSet())
+
+	stp.flagStaking.Toggle(epoch >= stp.unBondCorrectionEpoch)
```
Comment (SebastianMarian) [inline process/scToProtocol/stakingToPeer.go:419]: stp.flagUnBondCorrection

File: process/scToProtocol/stakingToPeer.go [hunk 6]
```diff
@@ -408,6 +415,9 @@ func (stp *stakingToPeer) getAllModifiedStates(body *block.Body) ([]string, erro
...
+	log.Debug("stakingToPeer: unBondCorrection", "enabled", stp.flagStaking.IsSet())
```
Comment (SebastianMarian) [inline process/scToProtocol/stakingToPeer.go:420]: stp.flagUnBondCorrection

File: process/scToProtocol/stakingToPeer.go (hunk 4 above)
Comment (SebastianMarian) [inline process/scToProtocol/stakingToPeer.go:57]: unBondCorrectionEnableEpoch

File: process/scToProtocol/stakingToPeer.go [hunk 7]
```diff
@@ -35,6 +35,7 @@ type ArgStakingToPeer struct {
 	CurrTxs          dataRetriever.TransactionCacher
 	RatingsData      process.RatingsInfoHandler
 	StakeEnableEpoch uint32
+	UnBondCorrection uint32
```
Comment (SebastianMarian) [inline process/scToProtocol/stakingToPeer.go:38]: UnBondCorrectionEnableEpoch

File: process/scToProtocol/stakingToPeer.go (hunk 5 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:419]: done

File: process/scToProtocol/stakingToPeer.go (hunk 6 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:420]: done

File: process/scToProtocol/stakingToPeer.go (hunk 4 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:57]: done

File: process/scToProtocol/stakingToPeer.go (hunk 7 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:38]: done

File: cmd/node/factory/structs.go (hunk 3 above)
Comment (sasurobert) [reply cmd/node/factory/structs.go:2078]: done

**Review by iulianpascalau — DISMISSED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**